schema-serializer/src/main/java/io/confluent/kafka/serializers/AbstractKafkaSchemaSerDe.java

```diff
@@ -467,24 +485,62 @@ protected Object executeRules(
         continue;
       }
       RuleContext ctx = new RuleContext(source, target,
-          subject, topic, headers, isKey, ruleMode, rule);
-      RuleExecutor ruleExecutor = ruleExecutors.get(rule.getType());
+          subject, topic, headers, original, isKey, ruleMode, rule);
+      RuleExecutor ruleExecutor = ruleExecutors.get(rule.getType().toUpperCase(Locale.ROOT));
       if (ruleExecutor == null) {
         log.warn("Could not find rule executor of type {}", rule.getType());
         return message;
       }
       try {
         message = ruleExecutor.transform(ctx, message);
-        if (message == null) {
-          throw new SerializationException("Validation failed for rule " + rule);
+        if (message != null) {
+          runAction(ctx, ruleMode, rule, rule.getOnSuccess(), message, null, null);
+        } else {
+          runAction(ctx, ruleMode, rule, rule.getOnFailure(), null, null, ErrorAction.TYPE);
         }
       } catch (RuleException e) {
-        throw new SerializationException("Could not execute rule " + rule, e);
+        runAction(ctx, ruleMode, rule, rule.getOnFailure(), message, e, ErrorAction.TYPE);
       }
     }
     return message;
   }
 
+  private void runAction(RuleContext ctx, RuleMode ruleMode, Rule rule, String action,
+      Object message, RuleException ex, String defaultAction) {
+    String actionName = getRuleActionName(rule, ruleMode, action);
+    if (actionName == null) {
+      actionName = defaultAction;
+    }
+    if (actionName != null) {
+      RuleAction ruleAction = ruleActions.get(actionName.toUpperCase(Locale.ROOT));
+      try {
+        ruleAction.run(ctx, message, ex);
+      } catch (RuleException e) {
+        log.error("Could not run post-rule action " + action, e);
+      }
+    }
+  }
+
+  private String getRuleActionName(Rule rule, RuleMode ruleMode, String actionName) {
+    if ((rule.getMode() == RuleMode.WRITEREAD || rule.getMode() == RuleMode.UPDOWN)
+        && actionName != null
+        && actionName.contains(",")) {
+      String[] parts = actionName.split(",");
+      switch (ruleMode) {
+        case WRITE:
+        case UPGRADE:
+          return parts[0];
+        case READ:
+        case DOWNGRADE:
+          return parts[0];
```

Comment:
Why are they both return `parts[0]`? Do you mean the user can put a common-sperate list of actions, but only the first one will be active?

Reply:
That's a copy-paste error, should be parts[1]